upup/pkg/fi/cloudup/awstasks/network_load_balancer.go

```diff
@@ -0,0 +1,855 @@
+			}
+
+			targetGroupArn = *response.TargetGroups[0].TargetGroupArn
+
+			if err := t.AddELBV2Tags(targetGroupArn, e.Tags); err != nil {
+				return err
+			}
+		}
+
+		{
+			for loadBalancerPort, listener := range e.Listeners {
+				loadBalancerPortInt, err := strconv.ParseInt(loadBalancerPort, 10, 64)
+				if err != nil {
+					return fmt.Errorf("error parsing load balancer listener port: %q", loadBalancerPort)
+				}
+				awsListener := listener.mapToAWS(loadBalancerPortInt, targetGroupArn, loadBalancerArn)
+
+				klog.V(2).Infof("Creating Listener for NLB")
+				_, err = t.Cloud.ELBV2().CreateListener(awsListener)
+				if err != nil {
+					return fmt.Errorf("Error creating listener for NLB: %v", err)
+				}
+			}
+		}
+	} else {
+		loadBalancerName = fi.StringValue(a.LoadBalancerName)
+
+		lb, err := findNetworkLoadBalancerByLoadBalancerName(t.Cloud, loadBalancerName)
+		if err != nil {
+			return fmt.Errorf("error getting load balancer by name: %v", err)
+		}
+
+		// if lb == nil {
+		// 	return fmt.Errorf("error querying nlb: %v", err)
+		// }
+
+		loadBalancerArn = *lb.LoadBalancerArn
+		tg, err := findTargetGroupByLoadBalancerArn(t.Cloud, loadBalancerArn)
+		if err != nil {
+			return fmt.Errorf("error getting target group by lb arn %v", loadBalancerArn)
+		}
+
+		targetGroupArn = *tg.TargetGroupArn
+
+		if changes.Subnets != nil {
+			var expectedSubnets []string
+			for _, s := range e.Subnets {
+				expectedSubnets = append(expectedSubnets, fi.StringValue(s.ID))
+			}
+
+			var actualSubnets []string
+			for _, s := range a.Subnets {
+				actualSubnets = append(actualSubnets, fi.StringValue(s.ID))
+			}
+
+			oldSubnetIDs := slice.GetUniqueStrings(expectedSubnets, actualSubnets)
+			if len(oldSubnetIDs) > 0 {
+				/*request := &elb.DetachLoadBalancerFromSubnetsInput{}
+				request.SetLoadBalancerName(loadBalancerName)
+				request.SetSubnets(aws.StringSlice(oldSubnetIDs))
+
+				klog.V(2).Infof("Detaching Load Balancer from old subnets")
+				if _, err := t.Cloud.ELB().DetachLoadBalancerFromSubnets(request); err != nil {
+					return fmt.Errorf("Error detaching Load Balancer from old subnets: %v", err)
+				}*/
+				//TODO: Seems likely we would have to delete and recreate in this case.
+				return fmt.Errorf("NLB's don't support detaching subnets, perhaps we need to recreate the NLB")
+			}
+
+			newSubnetIDs := slice.GetUniqueStrings(actualSubnets, expectedSubnets)
+			if len(newSubnetIDs) > 0 {
+
+				request := &elbv2.SetSubnetsInput{}
+				request.SetLoadBalancerArn(loadBalancerArn)
+				request.SetSubnets(aws.StringSlice(append(actualSubnets, newSubnetIDs...)))
+
+				klog.V(2).Infof("Attaching Load Balancer to new subnets")
+				if _, err := t.Cloud.ELBV2().SetSubnets(request); err != nil {
+					return fmt.Errorf("Error attaching Load Balancer to new subnets: %v", err)
+				}
+			}
+		}
+
+		//TODO: decide if security groups should be applied to master nodes
```

Comment:
We can remove all SG references in the NLB tasks. Rather than having the NLB tasks apply SGs to the masters the models will setup the master ASGs to update their SGs if we're using an NLB (which it looks like you already did in pkg/model/awsmodel/api_loadbalancer.go )

Reply:
I noticed this was being done @ load_balancer.go L581-592 and I am trying to sort out why thats needed there / what edge case that is for...and if thats valid, does that also apply in the NLB case as well..